statev2/src/replication/log_store.rs

```diff
@@ -0,0 +1,233 @@
+//! Defines the storage layer for the `raft` implementation. We store logs, snapshots,
+//! metadata, etc in the storage layer -- concretely an embedded KV store
+
+use std::sync::Arc;
+
+use libmdbx::{TransactionKind, RO};
+use protobuf::Message;
+use raft::{
+    prelude::{
+        ConfState, Entry as RaftEntry, HardState, Snapshot as RaftSnapshot, SnapshotMetadata,
+    },
+    Error as RaftError, GetEntriesContext, RaftState, Result as RaftResult, Storage,
+    StorageError as RaftStorageError,
+};
+
+use crate::storage::{
+    cursor::DbCursor,
+    db::{DbTxn, DB},
+    ProtoStorageWrapper,
+};
+
+use super::error::ReplicationError;
+
+// -------------
+// | Constants |
+// -------------
+
+/// The name of the raft metadata table in the database
+pub const RAFT_METADATA_TABLE: &str = "raft-metadata";
+/// The name of the raft logs table in the database
+pub const RAFT_LOGS_TABLE: &str = "raft-logs";
+
+/// The name of the raft hard state key in the KV store
+pub const HARD_STATE_KEY: &str = "hard-state";
+/// The name of the raft conf state key in the KV store
+pub const CONF_STATE_KEY: &str = "conf-state";
+/// The name of the snapshot metadata key in the KV store
+pub const SNAPSHOT_METADATA_KEY: &str = "snapshot-metadata";
+
+// -----------
+// | Helpers |
+// -----------
+
+/// Parse a raft LSN from a string
+fn parse_lsn(s: &str) -> Result<u64, ReplicationError> {
+    s.parse::<u64>()
+        .map_err(|_| ReplicationError::ParseValue(s.to_string()))
+}
+
+/// Format a raft LSN as a string
+fn lsn_to_key(lsn: u64) -> String {
+    lsn.to_string()
+}
+
+// -------------
+// | Log Store |
+// -------------
+
+/// The central storage abstraction, wraps a KV database
+pub struct LogStore {
+    /// The underlying database reference
+    db: Arc<DB>,
+}
+
+impl LogStore {
+    /// Constructor
+    pub fn new(db: Arc<DB>) -> Result<Self, ReplicationError> {
+        // Create the logs table in the db
+        db.create_table(RAFT_METADATA_TABLE)
```

Comment:
Should we have a way to get a handle on the `LogStore` in the runtime without creating the logs table? E.g. if recovering from a crash, the table should already exist, right?

Reply:
That's what `create_table` does, if the table already exists it's a no-op